### PR TITLE
Remove unused input 'num_batches_tracked' for batch norm from onnx model

### PR DESCRIPTION
**Pull request by neginraoof**
[WIP]
'num_batches_tracked' is exported through the state dictionary, to the onnx model as an input. However, because it is not used in the exported graph, it is considered as unused input. 
Currently ORT warns users about the unused input node in their model's graph.
Also, we currently don't support batch_norm with track_running_stats = False in our symbolic op for export. This is because setting track_running_stats = False will change the logic of batch_norm, and so running_mean and running_var won't be tracked (will show in None) at export. This behavior (in symboli_opset9) can change in future if there is demand for this flavor of batch_norm.

### CODE REVIEW

**Comment by ezyang** [comment]
cc @lara-hdr

**Comment by neginraoof** [comment]
@BowenBao please review

**Review by BowenBao — APPROVED** [review]
LGTM

**Review by lara-hdr — APPROVED** [review]
CI failure seem unrelated. Looks good.

**Comment by zhangguanheng66** [comment]
@houseroad Could you take a look the PR? Thanks.

**Comment by neginraoof** [comment]
cc @bddppq

**Review by spandantiwari — APPROVED** [review]
LGTM

File: torch/jit/__init__.py [hunk 1]
```diff
@@ -275,6 +275,17 @@ def _unique_state_dict(module, keep_vars=False):
     return filtered_dict
 
 
+def _remove_unused_state_batchnorm(module, state_dict):
```
Comment (bddppq) [inline torch/jit/__init__.py]: Looks to me this is onnx specific, could you move this to torch/onnx?

File: torch/onnx/symbolic_opset9.py [hunk 2]
```diff
@@ -873,6 +873,11 @@ def _convolution(g, input, weight, bias, stride, padding, dilation,
 
 @parse_args('v', 'v', 'v', 'v', 'v', 'i', 'f', 'f', 'i')
 def batch_norm(g, input, weight, bias, running_mean, running_var, training, momentum, eps, cudnn_enabled):
+    if (running_mean is None or running_mean.node().mustBeNone()) \
+            and (running_var is None or running_var.node().mustBeNone()):
+        return _unimplemented("batch_norm", "running_mean and running_var are null. "
```
Comment (bddppq) [inline torch/onnx/symbolic_opset9.py]: nit: `null` -> `None`

**Review by zou3519 — CHANGES_REQUESTED** [review]
@bddppq requested changes

File: torch/jit/__init__.py (hunk 1 above)
Comment (neginraoof) [reply torch/jit/__init__.py]: This is necessary in LegacyTracedModule as well. There are unused keys in the state dict, which should be removed.

Comment (neginraoof) [reply torch/jit/__init__.py]: @bddppq

Comment (neginraoof) [reply torch/jit/__init__.py]: @bddppq please review the updates

**Comment by neginraoof** [comment]
@bddppq for review

File: torch/jit/__init__.py [hunk 3]
```diff
@@ -273,6 +273,17 @@ def _unique_state_dict(module, keep_vars=False):
...
+    # Removing unused input 'num_batches_tracked' for BatchNorm. 'num_batches_tracked' is stored in
+    # register_buffer, but it's not used at ONNX export. In future, to preserve 'num_batches_tracked'
+    # in register_buffer for training, module's attribute 'training' could be used.
+    batchnorm = 'bn'
+    for sub_name, sub_mod in module.named_modules():
+        if batchnorm in sub_name:
+            state_dict.pop(sub_name + '.num_batches_tracked', None)
```
[the remark this answers is not in the document]
Comment (houseroad) [reply torch/jit/__init__.py]: Agreed with @dzhulgakov , hardcode is a bad practice. I also prefer adding a pass.

**Review by houseroad — CHANGES_REQUESTED** [review]
Please address the inline comments.

**Comment by houseroad** [comment]
friendly ping for update?

**Comment by houseroad** [comment]
friendly ping for update?

**Comment by neginraoof** [comment]
@houseroad Sorry I've been prioritizing ops. I'm looking into this now.

File: torch/csrc/jit/passes/onnx/peephole.cpp [hunk 4]
```diff
@@ -620,6 +620,25 @@ void removeMaxPoolUnusedOutput(Block* b) {
...
+void removeBatchNormUnusedInput(Block* b) {
+  for (auto it = b->nodes().begin(), end = b->nodes().end(); it != end; ++it) {
+    auto n = *it;
+    for (auto* child_block : n->blocks()) {
+      removeBatchNormUnusedInput(child_block);
+    }
+    if (strcmp(n->kind().toQualString(), "onnx::BatchNormalization") == 0) {
+      int ind = 0;
+      for (auto* input : n->inputs()) {
+        std::cout << input->debugName() << " has uses " << input->hasUses() << "\n";
```
Comment (houseroad) [inline torch/csrc/jit/passes/onnx/peephole.cpp]: remove debug information?

File: torch/csrc/jit/passes/onnx/peephole.cpp [hunk 5]
```diff
@@ -620,6 +620,25 @@ void removeMaxPoolUnusedOutput(Block* b) {
   }
 }
 
+void removeBatchNormUnusedInput(Block* b) {
```
Comment (houseroad) [inline torch/csrc/jit/passes/onnx/peephole.cpp]: how about generalizing this pass? We can have a white list of the operator names.

File: test/onnx/test_utility_funs.py [hunk 6]
```diff
@@ -223,6 +223,37 @@ def test_error_on_data_parallel(self):
...
+        class BatchNormModule(torch.nn.Module):
+            def __init__(self):
+                super(BatchNormModule, self).__init__()
+                self.bn = torch.nn.BatchNorm1d(224)
+
+            def forward(self, input):
+                return self.bn(input)
+
+        x = torch.randn(2, 224)
+        graph, params_dict, __ = utils._model_to_graph(BatchNormModule(), (x,))
```
Comment (houseroad) [inline test/onnx/test_utility_funs.py]: why __ instead _?

File: test/onnx/test_utility_funs.py [hunk 7]
```diff
@@ -223,6 +223,37 @@ def test_error_on_data_parallel(self):
...
+            def __init__(self):
+                super(BatchNormModule, self).__init__()
+                self.bn = torch.nn.BatchNorm1d(224, track_running_stats=False)
+
+            def forward(self, input):
+                return self.bn(input)
+
+        f = io.BytesIO()
+        x = torch.randn(2, 224)
+        graph, params_dict, __ = utils._model_to_graph(BatchNormModule(), (x,))
```
Comment (houseroad) [inline test/onnx/test_utility_funs.py]: ditto

File: test/onnx/test_utility_funs.py [hunk 8]
```diff
@@ -223,6 +223,37 @@ def test_error_on_data_parallel(self):
...
+        for input in graph.inputs():
+            assert len(input.uses()) > 0
+        with self.assertRaises(RuntimeError):
+            torch.onnx.export(BatchNormModule(), x, f, opset_version=self.opset_version)
```
Comment (houseroad) [inline test/onnx/test_utility_funs.py]: why do we expect runtime error?

**Review by houseroad — CHANGES_REQUESTED** [review]
Please see my inline comments.

**Comment by edurenye** [comment]
Why was it closed?
